doc/lspconfig.txt

```diff
@@ -1,242 +1,628 @@
-* `:LspStop <client_id>` Defaults to stopping all buffer clients.
-* `:LspRestart <client_id>` Defaults to restarting all buffer clients.
 ==============================================================================
-10. The wiki                                              *lspconfig-the-wiki*
+ROOT DETECTION                                        *lspconfig-root-detection*
+                                                            *lspconfig-root-dir*
+
+lspconfig automatically launches language servers by defining a filetype
+autocommand based on the `filetypes` specified in the default configuration of
+each server, optionally overridable by the `filetypes` table passed to
+`setup`.
+
+This autocommand triggers a search from the current file position in the
+filesystem hierarchy up to the top level directory of your filesystem. The
+`root_dir` entry of each configuration is a function that returns true if the
+current directory in this traversal matches a given root pattern.
+
+The following utility functions are provided by lspconfig. Each function call
+below, returns a function that takes as it's argument the current buffer path.
+Examples are provided for didactic purpose:
+
+- `util.root_pattern`: variadic function which takes multiple arguments, which
+  corresponding to a different root pattern against which the contents of the
+  current directory in the traversal will be matched against using
+  `vim.fn.glob`. See `:help vim.fn.glob` for details.
+>
+   root_dir = util.root_pattern('pyproject.toml', 'requirements.txt')
+<
+- `util.find_git_ancestor`: locates the first parent directory containing a
+  `.git` directory.
+>
+   root_dir = util.find_git_ancestor
 
+- `util.find_node_modules_ancestor`:  locates the first parent directory
+  containing a `node_modules` directory.
+>
+   root_dir = util.find_node_modules_ancestor
+<
 
-Please see the [wiki](https://github.com/neovim/nvim-lspconfig/wiki) for
-additional topics, including:
-* [Installing language servers automatically](https://github.com/neovim/nvim-lspconfig/wiki/Installing-language-servers-automatically)
-* [Snippets support](https://github.com/neovim/nvim-lspconfig/wiki/Snippets)
-* [Project local settings](https://github.com/neovim/nvim-lspconfig/wiki/Project-local-settings)
-* [Recommended plugins for enhanced language server features](https://github.com/neovim/nvim-lspconfig/wiki/Language-specific-plugins)
+- `util.find_package_json_ancestor`:  locates the first parent directory
+  containing a `package.json`.
+>
+   root_dir = util.find_json_ancestor
+<
 ==============================================================================
-11. Windows                                                *lspconfig-windows*
+COMPOSITIONAL ROOT DIRECTORIES                      *lspconfig-root-composition*
```

Comment:
-> "ADVANCED ROOT DIRECTORY DETECTION"

("Compositional" is awkward and not immediately understandable.)